src/D2L.CodeStyle.Analyzers/Pinning/PinnedAttributeAnalyzer.cs

```diff
@@ -0,0 +1,70 @@
+﻿using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace D2L.CodeStyle.Analyzers.Pinning {
+	[DiagnosticAnalyzer( LanguageNames.CSharp )]
+	internal sealed class PinnedAttributeAnalyzer : DiagnosticAnalyzer {
+
+		private static readonly TypeKind[] TypeKindsSupported = new[] { TypeKind.Class, TypeKind.Interface };
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(
+			Diagnostics.PinnedTypesMustNotMove
+			);
+
+		public override void Initialize( AnalysisContext context ) {
+			context.ConfigureGeneratedCodeAnalysis( GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics );
+			context.EnableConcurrentExecution();
+			context.RegisterCompilationStartAction( OnCompilationStart );
+		}
+
+		private void OnCompilationStart(
+			CompilationStartAnalysisContext context
+		) {
+			context.RegisterSymbolAction( AnalyzeSymbol,
+				SymbolKind.NamedType );
+		}
+
+		private static void AnalyzeSymbol( SymbolAnalysisContext context ) {
+			INamedTypeSymbol? pinnedAttributeSymbol = context.Compilation.GetTypeByMetadataName( PinnedAnalyzerHelper.PinnedAttributeName );
+			if( pinnedAttributeSymbol == null ) {
+				return;
+			}
+
+			INamedTypeSymbol? classSymbol = context.Symbol as INamedTypeSymbol;
+			if( classSymbol == null ) {
+				return;
+			}

```

Comment:
Consider hard-casting this rather than giving the option for a silent failure. This action only runs on `SymbolKind.NamedType` so a cast to `INamedTypeSymbol` should never fail, and if it ever does, we'd want to know very loudly rather than having this silently skip analysis.